src/actions/attack.cpp

```diff
@@ -1604,6 +1604,57 @@ int under_leadership(const unit &u, const map_location& loc, const_attack_ptr we
 	return leader_effect.get_composite_value();
 }
 
+static void alignment_list(unit_ability_list& abil_list)
+{	//check if have abilities or specials alignment and remove alignment bad specified
+	const std::set<std::string> checking_alignment = {"neutral", "lawful", "chaotic", "liminal"};
+	for(unit_ability_list::iterator i = abil_list.begin(); i != abil_list.end();) {
+		if(checking_alignment.count((*i->ability_cfg)["alignment"].str()) == 0){
+			i = abil_list.erase(i);
+		} else {
+			++i;
+		}
+	}
+}
+
+static std::pair<int, bool>alignment_bonus(unit_ability_list& abil_list, const std::string& alignment, int bonus_value)
+{
+	for(const auto& i : abil_list) {
+		if((*i.ability_cfg)["alignment"].str() == alignment) {
+			return {bonus_value, true};
+		}
+	}
+	return {0, false};
+}
+
+unit_alignments::type attack_type::specials_alignment() const
+{
+	unit_alignments::type alignment = (*self_).alignment();
+	unit_alignments::type temp_alignment = (*self_).alignment();
+	unit_ability_list abil_list = get_specials_and_abilities("attack_alignment");
+	alignment_list(abil_list);
+	if(abil_list.empty()){
+		return alignment;
+	}
+	bool is_liminal = alignment_bonus(abil_list, "liminal");
+	bool is_neutral = alignment_bonus(abil_list, "neutral");
+	bool is_lawful = alignment_bonus(abil_list, "lawful");
+	bool is_chaotic = alignment_bonus(abil_list, "chaotic");
+	std::string str_alignment;
+	if(is_liminal){str_alignment = "liminal";}
+	else if((is_neutral && !is_lawful && !is_chaotic) || (is_lawful && is_chaotic)){str_alignment = "neutral";}
+	else if( is_lawful){str_alignment = "lawful";}
+	else if( is_chaotic){str_alignment = "chaotic";}
+	auto new_align = unit_alignments::get_enum(str_alignment);
```

Comment:
As it's an enum, please use the enum in the lines above instead of using strings and then converting them.

I've been doing some local experiments, will push a branch with my work-in-progress today.